pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1322 @@
+								total_collators,
+								total_delegators,
+							));
+						}
+						None
+					}
+				})
+				.collect::<Vec<Bond<T::AccountId, RoundIndex>>>();
+			<ExitQueue<T>>::put(OrderedSet::from(remain_exits));
+		}
+
+		/// Select the top `n` collators in terms of cumulated stake (self +
+		/// from delegators) from the CandidatePool to become block authors for
+		/// the next round.
+		fn select_top_candidates(next: RoundIndex) -> (u32, BalanceOf<T>, BalanceOf<T>) {
+			let (mut all_collators, mut total_collators, mut total_delegators) =
+				(0u32, BalanceOf::<T>::zero(), BalanceOf::<T>::zero());
+			let mut candidates = <CandidatePool<T>>::get().0;
+
+			// Order candidates by their total stake
+			// TODO: Safe to unwrap?
+			candidates.sort_unstable_by(|a, b| a.amount.partial_cmp(&b.amount).unwrap());
+			let top_n = <TotalSelected<T>>::get() as usize;
+
+			// Choose the top TotalSelected qualified candidates, ordered by stake (least to
+			// greatest, thus requires `rev()`)
+			let mut collators = candidates
+				.into_iter()
+				.rev()
+				.take(top_n)
+				.filter(|x| x.amount >= T::MinCollatorStk::get())
+				.map(|x| x.owner)
+				.collect::<Vec<T::AccountId>>();
+
+			// Snapshot exposure for round for weighting reward distribution
+			for account in collators.iter() {
+				let state = <CollatorState<T>>::get(&account).expect("all members of CandidateQ must be candidates");
+				let amount_collator = state.bond;
+				let amount_delegators = state.total.saturating_sub(amount_collator);
+				let exposure: CollatorSnapshot<T::AccountId, BalanceOf<T>> = state.into();
+				<AtStake<T>>::insert(next, account, exposure);
+				all_collators += 1u32;
+				total_collators = total_collators.saturating_add(amount_collator);
+				total_delegators = total_delegators.saturating_add(amount_delegators);
+				Self::deposit_event(Event::CollatorChosen(
+					next,
+					account.clone(),
+					amount_collator,
+					amount_delegators,
+				));
+			}
+			collators.sort();
+
+			// Insert canonical collator set
+			<SelectedCandidates<T>>::put(collators);
+			(all_collators, total_collators, total_delegators)
+		}
+
+		fn update_inflation(inflation: InflationInfo) -> Result<(), DispatchError> {
+			ensure!(inflation.is_valid(), Error::<T>::InvalidSchedule);
+			Self::deposit_event(Event::RoundInflationSet(
+				inflation.collator.max_rate,
+				inflation.collator.reward_rate.round,
+				inflation.delegator.max_rate,
+				inflation.delegator.reward_rate.round,
+			));
+			<InflationConfig<T>>::put(inflation);
+			Ok(())
+		}
+
+		// Weight: writes(2)
+		fn do_reward(who: &T::AccountId, reward: BalanceOf<T>, now: T::BlockNumber) {
+			// mint
+			if reward > T::Currency::minimum_balance() {
+				if let Ok(imb) = T::Currency::deposit_into_existing(who, reward) {
+					// TODO: Remove?
+					Self::deposit_event(Event::Rewarded(who.clone(), imb.peek()));
+				}
+			}
+
+			// set & update lock
+			let mut locks = <RewardLocks<T>>::get(who);
```

Comment:
Bug.

This gets called even if the above if statement would not deposit a reward.